crates/revm/revm-inspectors/src/tracing/builder/parity.rs

```diff
@@ -218,13 +238,153 @@ impl ParityTraceBuilder {
+
+                            current = self.nodes.get(parent).expect("there should be a parent");
+                        }
+                        None => break instructions,
+                    }
+                }
+            }
+        };
+
+        VmTrace { code: Default::default(), ops: instructions }
+    }
+
+    /// todo::n config
+    ///
+    /// Creates a VM instruction from a [CallTraceStep] and a [VmTrace] for the subcall if there is one
+    fn make_instruction(step: &CallTraceStep, maybe_sub: Option<VmTrace>) -> VmInstruction {
+        let maybe_storage = match step.storage_change {
+            Some(storage_change) => {
+                Some(StorageDelta { key: storage_change.key, val: storage_change.value })
+            }
+            None => None,
+        };
+
+        let maybe_memory = match step.memory.len() {
+            0 => None,
+            _ => {
+                Some(MemoryDelta { off: step.memory_size, data: step.memory.data().clone().into() })
+            }
+        };
+
+        let maybe_execution = Some(VmExecutedOperation {
+            used: step.gas_cost,
+            push: match step.new_stack {
+                Some(new_stack) => Some(new_stack.into()),
+                None => None,
+            },
+            mem: maybe_memory,
+            store: maybe_storage,
+        });
+
+        VmInstruction {
+            pc: step.pc,
+            cost: 0, // todo::n
+            ex: maybe_execution,
+            sub: maybe_sub,
+        }
+    }
 }
 
-/// Construct the vmtrace for the entire callgraph
-fn vm_trace(nodes: &[CallTraceNode]) -> VmTrace {
-    // TODO: populate vm trace
+/// addresses are presorted via breadth first walk thru [CallTraceNode]s, this  can be done by a walker in [crate::tracing::builder::walker]
+///
+/// iteratively fill the [VmTrace] code fields
+pub(crate) fn populate_vm_trace_bytecodes<DB, I>(
+    db: &DB,
+    breadth_first_addresses: I,
+    trace: &mut VmTrace,
+) -> Result<(), DB::Error>
+where
+    DB: DatabaseRef,
+    I: IntoIterator<Item = Address>,
+{
+    let mut stack: VecDeque<&mut VmTrace> = VecDeque::new();
+    stack.push_back(trace);
+
+    let mut addrs = breadth_first_addresses.into_iter();
+
+    loop {
+        match stack.pop_front() {
+            Some(curr_ref) => {
+                for op in curr_ref.ops.iter_mut() {
+                    if let Some(sub) = op.sub.as_mut() {
+                        stack.push_back(sub);
+                    }
+                }
+
+                let addr = addrs.next().expect("there should be an address");
 
-    VmTrace { code: nodes[0].trace.data.clone().into(), ops: vec![] }
+                let db_acc = db.basic(addr)?.unwrap_or_default();
+
+                curr_ref.code = match db_acc.code {
+                    Some(code) => code.bytecode.into(),
+                    None => Default::default(),
+                };
```

Comment:
I think we need to use `code_by_hash` here if the code_hash is __not__ KECCAK_EMPTY and code is none

because code is stored separately

Reply:
fixed